scanner/indexer/indexer.go

```diff
@@ -57,15 +73,146 @@ func NewIndexer(ctx context.Context) (*Indexer, error) {
 
 	indexer, err := libindex.New(ctx, &opts, c)
 	if err != nil {
-		return nil, errors.Wrap(err, "creating libindex")
+		return nil, fmt.Errorf("creating libindex: %w", err)
 	}
 
-	return &Indexer{
+	return &indexerImpl{
 		indexer: indexer,
 	}, nil
 }
 
 // Close closes the indexer.
-func (i *Indexer) Close(ctx context.Context) error {
+func (i *indexerImpl) Close(ctx context.Context) error {
 	return i.indexer.Close(ctx)
 }
+
+// IndexContainerImage creates a ClairCore index report for a given container
+// image. The manifest is populated with layers from the image specified by a
+// URL. This method performs a partial content request on each layer to generate
+// the layer's URI and headers.
+func (i *indexerImpl) IndexContainerImage(
+	ctx context.Context,
+	manifestDigest claircore.Digest,
+	imageURL string,
+	opts ...Option,
+) (*claircore.IndexReport, error) {
+	ctx = zlog.ContextWithValues(ctx, "component", "scanner/backend/indexer")
+	o := makeOptions(opts...)
+	imgRef, err := parseContainerImageURL(imageURL)
+	if err != nil {
+		return nil, fmt.Errorf("parsing image URL: %w", err)
+	}
+	imgLayers, err := getContainerImageLayers(ctx, o, imgRef)
+	if err != nil {
+		return nil, fmt.Errorf("listing image layers (reference %q): %w", imgRef.String(), err)
+	}
+	httpClient := http.Client{Timeout: time.Duration(1) * time.Minute}
+	manifest := &claircore.Manifest{
+		Hash: manifestDigest,
+	}
+	zlog.Info(ctx).
+		Str("image_reference", imgRef.String()).
+		Str("registry", imgRef.Context().RegistryStr()).
+		Int("layers_count", len(imgLayers)).
+		Msg("retrieving layers to populate container image manifest")
+	for _, layer := range imgLayers {
+		ccDigest, layerDigest, err := getLayerDigests(layer)
+		if err != nil {
+			return nil, fmt.Errorf("getting layer digests: %w", err)
+		}
+		// TODO Check for non-retriable errors (permission denied, etc.) to report properly.
+		layerReq, err := getLayerRequest(&httpClient, imgRef, layerDigest)
+		if err != nil {
+			return nil, fmt.Errorf("getting layer request URL and headers (digest: %q): %w",
+				layerDigest.String(), err)
+		}
+		layerReq.Header.Del("User-Agent")
+		layerReq.Header.Del("Range")
+		manifest.Layers = append(manifest.Layers, &claircore.Layer{
+			Hash:    ccDigest,
+			URI:     layerReq.URL.String(),
+			Headers: layerReq.Header,
+		})
+	}
+	return i.indexer.Index(ctx, manifest)
+}
+
+// getLayerDigests returns the clairclore and containerregistry digests for the layer.
```

Comment:
ClairCore

Reply:
That's a reference to the golang's package reference, hence all lower case.